modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/DestinationFactory.java

```diff
@@ -0,0 +1,117 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.sql.engine.exec;
+
+import static org.apache.ignite.internal.util.CollectionUtils.first;
+import static org.apache.ignite.internal.util.CollectionUtils.nullOrEmpty;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import org.apache.calcite.util.ImmutableIntList;
+import org.apache.ignite.internal.sql.engine.metadata.ColocationGroup;
+import org.apache.ignite.internal.sql.engine.metadata.NodeWithTerm;
+import org.apache.ignite.internal.sql.engine.schema.TableDescriptor;
+import org.apache.ignite.internal.sql.engine.trait.AllNodes;
+import org.apache.ignite.internal.sql.engine.trait.Destination;
+import org.apache.ignite.internal.sql.engine.trait.DistributionFunction;
+import org.apache.ignite.internal.sql.engine.trait.DistributionFunction.AffinityDistribution;
+import org.apache.ignite.internal.sql.engine.trait.IgniteDistribution;
+import org.apache.ignite.internal.sql.engine.trait.Partitioned;
+import org.apache.ignite.internal.sql.engine.trait.RandomNode;
+import org.apache.ignite.internal.sql.engine.util.Commons;
+import org.apache.ignite.internal.sql.engine.util.HashFunctionFactory;
+import org.apache.ignite.internal.util.IgniteUtils;
+
+/**
+ * Factory that resolves {@link IgniteDistribution} trait, which represents logical {@link DistributionFunction} function, into its
+ * physical representation - {@link Destination} function.
+ */
+public class DestinationFactory<RowT> {
+    private final HashFunctionFactory<RowT> hashFunctionFactory;
+    private final ResolvedDependencies dependencies;
+
+    /**
+     * Constructor.
+     *
+     * @param hashFunctionFactory Hash-function factory required to resolve hash-based distributions.
+     * @param dependencies Dependencies required to resolve row value dependent distributions.
+     */
+    DestinationFactory(HashFunctionFactory<RowT> hashFunctionFactory, ResolvedDependencies dependencies) {
+        this.hashFunctionFactory = hashFunctionFactory;
+        this.dependencies = dependencies;
+    }
+
+    /**
+     * Creates a destination based on given distribution and nodes mapping.
+     *
+     * @param distribution Distribution function.
+     * @param group Target mapping.
+     * @return Destination function.
+     */
+    Destination<RowT> createDestination(IgniteDistribution distribution, ColocationGroup group) {
+        DistributionFunction function = distribution.function();
+        ImmutableIntList keys = distribution.getKeys();
+
+        switch (function.type()) {
+            case SINGLETON:
+                if (group == null || group.nodeNames() == null || group.nodeNames().size() != 1) {
```

Comment:
I think it would be better to move argument checks of each destination type to appropriate constructor so that it would not be possible create an invalid destination. There is no penalty in doing so, since there is only one place where these destination are used they do not cross serialisation boundaries.

Reply:
converted to asserts